### ISSUE TITLE
Add a composite index on ping country + created_at

### ISSUE THREAD

**Issue by bochecha** [endlessm/azafea#26]
@ramcq tried to run the following query:

```sql
SELECT DISTINCT p.country,
                (SELECT count(pq1.id)
                   FROM ping_v1 pq1
                   WHERE pq1.country = p.country
                   AND pq1.created_at >= '2019-01-01'::date
                   AND pq1.created_at < '2019-04-01'::date) AS q1,
                (SELECT count(pq2.id)
                   FROM ping_v1 pq2
                   WHERE pq2.country = p.country
                   AND pq2.created_at >= '2019-04-01'::date
                   AND pq2.created_at < '2019-07-01'::date) AS q2
  FROM ping_v1 p
  WHERE p.created_at >= (now() - '1 day'::interval);
```

And it was very slow. The query plan looks like this:
```
                                                        QUERY PLAN                                                        
--------------------------------------------------------------------------------------------------------------------------
 Unique  (cost=4780.17..4782.27 rows=144 width=32)
   ->  Sort  (cost=4780.17..4780.70 rows=210 width=32)
         Sort Key: p.country, ((SubPlan 1)), ((SubPlan 2))
         ->  Bitmap Heap Scan on ping_v1 p  (cost=5.78..4772.07 rows=210 width=32)
               Recheck Cond: (created_at >= (now() - '1 day'::interval))
               ->  Bitmap Index Scan on ix_ping_v1_created_at  (cost=0.00..5.73 rows=210 width=0)
                     Index Cond: (created_at >= (now() - '1 day'::interval))
               SubPlan 1
                 ->  Aggregate  (cost=11.31..11.32 rows=1 width=8)
                       ->  Bitmap Heap Scan on ping_v1 pq1  (cost=4.18..11.30 rows=1 width=4)
                             Recheck Cond: ((created_at >= '2019-01-01'::date) AND (created_at < '2019-04-01'::date))
                             Filter: ((country)::text = (p.country)::text)
                             ->  Bitmap Index Scan on ix_ping_v1_created_at  (cost=0.00..4.18 rows=3 width=0)
                                   Index Cond: ((created_at >= '2019-01-01'::date) AND (created_at < '2019-04-01'::date))
               SubPlan 2
                 ->  Aggregate  (cost=11.31..11.32 rows=1 width=8)
                       ->  Bitmap Heap Scan on ping_v1 pq2  (cost=4.18..11.30 rows=1 width=4)
                             Recheck Cond: ((created_at >= '2019-04-01'::date) AND (created_at < '2019-07-01'::date))
                             Filter: ((country)::text = (p.country)::text)
                             ->  Bitmap Index Scan on ix_ping_v1_created_at  (cost=0.00..4.18 rows=3 width=0)
                                   Index Cond: ((created_at >= '2019-04-01'::date) AND (created_at < '2019-07-01'::date))
```

Adding an index on `country` does not have any significant impact.

However, adding a composite index on `country` and `created_at`, the query plan becomes:
```
                                                                          QUERY PLAN                                                                          
--------------------------------------------------------------------------------------------------------------------------------------------------------------
 Unique  (cost=3465.26..3467.36 rows=144 width=32)
   ->  Sort  (cost=3465.26..3465.78 rows=210 width=32)
         Sort Key: p.country, ((SubPlan 1)), ((SubPlan 2))
         ->  Bitmap Heap Scan on ping_v1 p  (cost=5.78..3457.16 rows=210 width=32)
               Recheck Cond: (created_at >= (now() - '1 day'::interval))
               ->  Bitmap Index Scan on ix_ping_v1_created_at  (cost=0.00..5.73 rows=210 width=0)
                     Index Cond: (created_at >= (now() - '1 day'::interval))
               SubPlan 1
                 ->  Aggregate  (cost=8.17..8.18 rows=1 width=8)
                       ->  Index Scan using ix_ping_v1_country_created_at on ping_v1 pq1  (cost=0.15..8.17 rows=1 width=4)
                             Index Cond: (((country)::text = (p.country)::text) AND (created_at >= '2019-01-01'::date) AND (created_at < '2019-04-01'::date))
               SubPlan 2
                 ->  Aggregate  (cost=8.17..8.18 rows=1 width=8)
                       ->  Index Scan using ix_ping_v1_country_created_at on ping_v1 pq2  (cost=0.15..8.17 rows=1 width=4)
                             Index Cond: (((country)::text = (p.country)::text) AND (created_at >= '2019-04-01'::date) AND (created_at < '2019-07-01'::date))
```

The query still has a high cost, but it's roughly 27% better.

This will require implementing #4 first.

**Comment by liZe** [comment]
The main problem with this request is not the index, it’s the request itself.

My original guess was that the subrequests were executed once per country, which is really bad. It’s actually even worst: it’s executed once per raw matching the last `created_at` filter. Adding an index on `country` + `created_at` improves the speed (`EXPLAIN` cost is divided by 5), but it’s not enough.

The same request can be written as:
```sql
SELECT q0.country, q1.count, q2.count
FROM (
  SELECT DISTINCT country
  FROM ping_v1 pq0
  WHERE pq0.created_at >= (now() - '1 day'::interval)
) AS q0
LEFT OUTER JOIN (
  SELECT country, count(pq1.id) AS count
  FROM ping_v1 pq1
  WHERE pq1.created_at >= '2019-01-01'
  AND pq1.created_at < '2019-04-01'
  GROUP BY country
) AS q1
USING (country)
LEFT OUTER JOIN (
  SELECT country, count(pq2.id) AS count
  FROM ping_v1 pq2
  WHERE pq2.created_at >= '2019-04-01'
  AND pq2.created_at < '2019-07-01'
  GROUP BY country
) AS q2
USING (country);
```

The only difference I can spot in the result is that we have `NULL` instead of `0` when there’s no country ping during a given quarter. The cost is divided by 20,000 (!) on the production database, we get the result in ~30s.

It’s possible to get even faster results, using an index on the date part of `created_at`:
```sql
CREATE INDEX ON ping_v1 (date_trunc('day', timezone('UTC', created_at)));
```

Using this new request

```sql
SELECT q0.country, q1.count, q2.count
FROM (
  SELECT DISTINCT country
  FROM ping_v1 pq0
  WHERE pq0.created_at >= (now() - '1 day'::interval)
) AS q0
LEFT OUTER JOIN (
  SELECT country, count(pq1.id) AS count
  FROM ping_v1 pq1
  WHERE date_trunc('day', timezone('UTC', pq1.created_at)) >= '2019-01-01'
  AND date_trunc('day', timezone('UTC', pq1.created_at)) < '2019-04-01'
  GROUP BY country
) AS q1
USING (country)
LEFT OUTER JOIN (
  SELECT country, count(pq2.id) AS count
  FROM ping_v1 pq2
  WHERE date_trunc('day', timezone('UTC', pq2.created_at)) >= '2019-04-01'
  AND date_trunc('day', timezone('UTC', pq2.created_at)) < '2019-07-01'
  GROUP BY country
) AS q2
USING (country);
```

gives an extra ×15 speed (~2s).

The problem with this kind of indexes is that we need to explicitly cast `created_at` each time we use it, and that this cast has to include the timezone information because indexes need to be immutable (truncating a date with no explicit timezone gives different results depending on the locale timezone).

**Comment by wjt** [comment]
I'm not @ramcq but I can guess at why he ran this query. Suppose that we made a big investment in promoting Endless OS in some particular countries in Q1 2019, and then reduced this in Q2 2019. We would then want to see how the usage of Endless OS changed in those countries between the two quarters – do we retain the users we acquired in Q1 through Q2, or did they fall off a cliff the moment we stopped promoting it?

Is there some way we can avoid the cast? Ping timestamps are always in server time (TTBOMK) so the timezone should (I hope!) always be UTC. I do remember seeing some case where Metabase's query builder interface caused casts to be inserted that meant that indexes couldn't be used.

More generally, you can imagine wanting to graph the number of pings per day per country. For what it's worth, this seems to be perfectly fast enough in my testing. Here's one example, charting daily pings per country for a particular image personality: https://metabase.endlessm.com/question/64. It's effectively instant. Maybe this is a non-issue.

**Comment by liZe** [comment]
> I'm not @ramcq but I can guess at why he ran this query. Suppose that we made a big investment in promoting Endless OS in some particular countries in Q1 2019, and then reduced this in Q2 2019. We would then want to see how the usage of Endless OS changed in those countries between the two quarters – do we retain the users we acquired in Q1 through Q2, or did they fall off a cliff the moment we stopped promoting it?

Thanks for the explanation. I suppose then that it’s better to have a clean request that we can easily edit, even if we have to wait 30 seconds to get the graph.

> Is there some way we can avoid the cast? Ping timestamps are always in server time (TTBOMK) so the timezone should (I hope!) always be UTC. I do remember seeing some case where Metabase's query builder interface caused casts to be inserted that meant that indexes couldn't be used.

I’m not sure that we can, at least easily. We have two distinct problems:
- PostgreSQL needs immutable indexes. Transforming a timestamp+timezone object into an abstract date (without tz, by definition) needs a reference timezone, and there’s logically no way we can avoid that. If we don’t set the reference timezone explicitly, the locale timezone is used and the index is not immutable anymore. I think that the latest ×15 speed enhancement requires this cast.
- When Metabase automatically casts timestamp+timezone columns, it doesn’t set an explicit timezone. So, this column will never benefit from an index, as this value depends on the server timezone and thus isn’t immutable.

**TL;DR**

Let’s forget the index+cast solution.

We can keep the first request, it doesn’t require any changes on the model. I didn’t find a way to write it using Metatable’s graphical editor, but SQL is easy to read and to edit. It takes 30 seconds on my backup database and less than 10 seconds on Metatable. It can give nice graphs, where you can easily find which countries have good or bad evolutions between quarters:

![Capture d’écran de 2020-07-10 13-53-39](https://user-images.githubusercontent.com/952243/87151854-d912a100-c2b4-11ea-91b1-162f0b8ebd42.png)
![Capture d’écran de 2020-07-10 14-05-12](https://user-images.githubusercontent.com/952243/87152679-71f5ec00-c2b6-11ea-80c2-2827713e7916.png)

And I’m sure that real users will be able to build even better grahps. When 2-letter country codes are in production, I think that we will be able to build a world map where countries are colored according to the evolution of their number of pings between months, quarters or years.

**Comment by liZe** [comment]
(That’s another problem also related to T29220 and metabase/pull#13081)

**Comment by liZe** [comment]
@wjt I think that this problem is fixed by the patch in metabase/metabase#13081, that’s applied on production (as far as I can remember).

**Comment by wjt** [comment]
Yep, think so!

**Comment by wjt** [comment]
I tried recreating your query above and mapping the difference over time.

https://metabase.endlessm.com/question/119

One interesting thing is that the query plan looks very different depending on whether I compare the last 30 days to the preceeding 30 days, or the last 90 days to the preceeding 90 days. 90 days uses a seq scan and takes 25 seconds:

```
azafea=> explain analyze SELECT
    q0.country,
    COALESCE(q1.count, 0) as earlier,
    COALESCE(q2.count, 0) as later,
    COALESCE(q2.count, 0) - COALESCE(q1.count, 0) as delta
FROM (
  SELECT DISTINCT country
  FROM ping_v1 pq0
  WHERE pq0.created_at >= (now() - '1 day'::interval)
) AS q0
LEFT OUTER JOIN (
  SELECT country, count(pq1.id) AS count
  FROM ping_v1 pq1
  WHERE pq1.created_at between now() - '180 days' :: interval and now() - '90 days' :: interval
  GROUP BY country
) AS q1
USING (country)
LEFT OUTER JOIN (
  SELECT country, count(pq2.id) AS count
  FROM ping_v1 pq2
  WHERE pq2.created_at between now() - '90 days' :: interval and now()
  GROUP BY country
) AS q2
USING (country);
                                                                               QUERY PLAN                                                                               
------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Hash Left Join  (cost=2092361.04..2092364.67 rows=144 width=27) (actual time=23433.595..23433.799 rows=152 loops=1)
   Hash Cond: ((pq0.country)::text = (q2.country)::text)
   ->  Merge Left Join  (cost=1112130.95..1112133.83 rows=144 width=11) (actual time=12394.907..12395.053 rows=152 loops=1)
         Merge Cond: ((pq0.country)::text = (q1.country)::text)
         ->  Sort  (cost=57726.62..57726.98 rows=144 width=3) (actual time=24.203..24.217 rows=152 loops=1)
               Sort Key: pq0.country
               Sort Method: quicksort  Memory: 32kB
               ->  HashAggregate  (cost=57718.58..57720.02 rows=144 width=3) (actual time=23.601..23.620 rows=152 loops=1)
                     Group Key: pq0.country
                     ->  Index Scan using ix_ping_v1_created_at on ping_v1 pq0  (cost=0.57..57669.78 rows=19521 width=3) (actual time=0.012..10.548 rows=42111 loops=1)
                           Index Cond: (created_at >= (now() - '1 day'::interval))
         ->  Sort  (cost=1054404.32..1054404.68 rows=144 width=11) (actual time=12370.699..12370.714 rows=196 loops=1)
               Sort Key: q1.country
               Sort Method: quicksort  Memory: 34kB
               ->  Subquery Scan on q1  (cost=1054396.28..1054399.16 rows=144 width=11) (actual time=12370.338..12370.405 rows=197 loops=1)
                     ->  HashAggregate  (cost=1054396.28..1054397.72 rows=144 width=11) (actual time=12370.336..12370.374 rows=197 loops=1)
                           Group Key: pq1.country
                           ->  Seq Scan on ping_v1 pq1  (cost=0.00..1039052.18 rows=3068821 width=7) (actual time=0.505..11525.961 rows=3024866 loops=1)
                                 Filter: ((created_at >= (now() - '180 days'::interval)) AND (created_at <= (now() - '90 days'::interval)))
                                 Rows Removed by Filter: 26872958
   ->  Hash  (cost=980228.30..980228.30 rows=144 width=11) (actual time=11038.661..11038.661 rows=188 loops=1)
         Buckets: 1024  Batches: 1  Memory Usage: 16kB
         ->  Subquery Scan on q2  (cost=980225.42..980228.30 rows=144 width=11) (actual time=11038.544..11038.605 rows=189 loops=1)
               ->  HashAggregate  (cost=980225.42..980226.86 rows=144 width=11) (actual time=11038.543..11038.573 rows=189 loops=1)
                     Group Key: pq2.country
                     ->  Seq Scan on ping_v1 pq2  (cost=0.00..964336.86 rows=3177712 width=7) (actual time=4301.824..10099.443 rows=3463251 loops=1)
                           Filter: ((created_at <= now()) AND (created_at >= (now() - '90 days'::interval)))
                           Rows Removed by Filter: 26434573
 Planning Time: 0.311 ms
 Execution Time: 23433.911 ms
(30 rows)
```

while 30 days hits the index and takes 1 second:

```
azafea=> explain analyze SELECT
    q0.country,
    COALESCE(q1.count, 0) as earlier,
    COALESCE(q2.count, 0) as later,
    COALESCE(q2.count, 0) - COALESCE(q1.count, 0) as delta
FROM (
  SELECT DISTINCT country
  FROM ping_v1 pq0
  WHERE pq0.created_at >= (now() - '1 day'::interval)
) AS q0
LEFT OUTER JOIN (
  SELECT country, count(pq1.id) AS count
  FROM ping_v1 pq1
  WHERE pq1.created_at between now() - '60 days' :: interval and now() - '30 days' :: interval
  GROUP BY country
) AS q1
USING (country)
LEFT OUTER JOIN (
  SELECT country, count(pq2.id) AS count
  FROM ping_v1 pq2
  WHERE pq2.created_at between now() - '30 days' :: interval and now()
  GROUP BY country
) AS q2
USING (country);
                                                                                  QUERY PLAN                                                                                  
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Hash Left Join  (cost=1980534.72..1980538.35 rows=144 width=27) (actual time=1170.081..1170.271 rows=152 loops=1)
   Hash Cond: ((pq0.country)::text = (q2.country)::text)
   ->  Merge Left Join  (cost=1056601.99..1056604.87 rows=144 width=11) (actual time=623.096..623.228 rows=152 loops=1)
         Merge Cond: ((pq0.country)::text = (q1.country)::text)
         ->  Sort  (cost=57726.62..57726.98 rows=144 width=3) (actual time=19.832..19.846 rows=152 loops=1)
               Sort Key: pq0.country
               Sort Method: quicksort  Memory: 32kB
               ->  HashAggregate  (cost=57718.58..57720.02 rows=144 width=3) (actual time=19.635..19.654 rows=152 loops=1)
                     Group Key: pq0.country
                     ->  Index Scan using ix_ping_v1_created_at on ping_v1 pq0  (cost=0.57..57669.78 rows=19521 width=3) (actual time=0.013..10.110 rows=42118 loops=1)
                           Index Cond: (created_at >= (now() - '1 day'::interval))
         ->  Sort  (cost=998875.37..998875.73 rows=144 width=11) (actual time=603.257..603.270 rows=177 loops=1)
               Sort Key: q1.country
               Sort Method: quicksort  Memory: 33kB
               ->  Subquery Scan on q1  (cost=998867.33..998870.21 rows=144 width=11) (actual time=602.968..603.022 rows=178 loops=1)
                     ->  HashAggregate  (cost=998867.33..998868.77 rows=144 width=11) (actual time=602.966..602.995 rows=178 loops=1)
                           Group Key: pq1.country
                           ->  Bitmap Heap Scan on ping_v1 pq1  (cost=34628.54..993066.37 rows=1160192 width=7) (actual time=85.660..258.637 rows=1139090 loops=1)
                                 Recheck Cond: ((created_at >= (now() - '60 days'::interval)) AND (created_at <= (now() - '30 days'::interval)))
                                 Heap Blocks: exact=30081
                                 ->  Bitmap Index Scan on ix_ping_v1_created_at  (cost=0.00..34338.49 rows=1160192 width=0) (actual time=80.971..80.971 rows=1139090 loops=1)
                                       Index Cond: ((created_at >= (now() - '60 days'::interval)) AND (created_at <= (now() - '30 days'::interval)))
   ->  Hash  (cost=923930.93..923930.93 rows=144 width=11) (actual time=546.957..546.957 rows=175 loops=1)
         Buckets: 1024  Batches: 1  Memory Usage: 16kB
         ->  Subquery Scan on q2  (cost=923928.05..923930.93 rows=144 width=11) (actual time=546.864..546.918 rows=176 loops=1)
               ->  HashAggregate  (cost=923928.05..923929.49 rows=144 width=11) (actual time=546.863..546.891 rows=176 loops=1)
                     Group Key: pq2.country
                     ->  Bitmap Heap Scan on ping_v1 pq2  (cost=28011.76..919235.76 rows=938458 width=7) (actual time=83.086..206.001 rows=1213636 loops=1)
                           Recheck Cond: ((created_at >= (now() - '30 days'::interval)) AND (created_at <= now()))
                           Heap Blocks: exact=11427
                           ->  Bitmap Index Scan on ix_ping_v1_created_at  (cost=0.00..27777.15 rows=938458 width=0) (actual time=81.643..81.643 rows=1213636 loops=1)
                                 Index Cond: ((created_at >= (now() - '30 days'::interval)) AND (created_at <= now()))
 Planning Time: 0.305 ms
 Execution Time: 1171.057 ms
(34 rows)
```

Not the first time I've seen this. I'm not sure how we can convince PostgreSQL to use the index in more cases...
